### ISSUE TITLE
Slam for openmvg

### ISSUE THREAD

**Issue by mojovski** [openMVG/openMVG#695]
It is desirable to add real time slam to openmvg.

1. definition of interfaces.
Input: a) images with cameracalibration
b) images without camera calibration
Output: ??? TBD

2. Fulfil the interface with orbslam or dvo codebase. Enables to evaluate, which is "better" for openmvg

3. Modular integration of orb, dvo, svo or other visual odometry/slam framework.

**Comment by pmoulon** [comment]
Thank you @mojovski for your enthusiasm.

Since most of the existing project license are contaminating, rather than fulfill the interface with existing projects it would be better to write our own or rewrite a pipeline. (See all the GPL listed here https://github.com/openMVG/awesome_3DReconstruction_list#opensource-slam)

We can start with a simple VO and then move to SLAM slowly.

As you said we must first list the interfaces that we be required and the new abstract concepts that will be required (relative pose graph...).

**Comment by waqarsqureshi** [comment]
So, the idea is to support camera post estimation is?
1 - Input video is processed to give visual odometry and then further interfaced with existing algorithm to get camera poses?

Would not the ORB license be good enough to interface with?

**Comment by pmoulon** [comment]
ORB SLAM license is GPLv3

**Comment by mojovski** [comment]
Just found this one. Any use?
https://github.com/avisingh599/mono-vo

**Comment by csp256** [comment]
@komrad36 and I should definitely be able to provide a better vo pipeline. Should we assume a calibrated camera? How important is modularity and customizability vs performance? Are we assuming single actor mono slam? Is global map consistency more important than local precision? Etc; what are our parameters?

**Comment by mojovski** [comment]
@csp256 , what type of customizability do you mean? Camera model? Math framework backend? Interchangeable festures (detector, decsriptor, matcher)?
I would propose to do easy evaluations and post some justifications for the selected variant. I mean, if orb behaves better than brist, then this is the choise and can be "burned" into the code.

If we start to talk about map consistency, things get messy. I would suggest to keep with VO for now. 

@pmoulon do you prefer this as a module inside openMvg or as a module as an external github project? :) In the latter case, extensions to map matching, loop closure, landmarks etc make more sense.

**Comment by pmoulon** [comment]
@csp256, @komrad36 I would be happy to provide help to work on working on a vo pipeline.
I think we can assume to use:
- a calibrated camera (or almost calibrated => we can run some BA to improve the intrinsics if the user ask it)
- we can consider having a first draft version and then working on improving speed and customizability.
   - I mean that abstraction is good to provide knowledge for people.
- we can first consider classic VO and then add a module to perform global map consistency (loop closure search and refinement)

@mojovski I would prefer working inside OpenMVG to make the project maintained as a core brick later on.

**Comment by waqarsqureshi** [comment]
So, the idea is to support camera post estimation is?
1 - Input video is processed to give visual odometry and then further interfaced with existing algorithm to get camera poses?

Would not the ORB license be good enough to interface with?

**Comment by waqarsqureshi** [comment]
Features could be a parameter as in MVG, such as SIFT, ORB, AKAZE or BRIEF?

**Comment by pmoulon** [comment]
@mojovski 
Any progress on that.
It would be nice to have but we need more ressources (contributors) to deal with such a project

**Comment by mojovski** [comment]
right, I feel that on my side it will take another months before I can deal with slam again...
recently I saw some code for visual odometry in OpenMVG. How does this relate to slam? Can we rely on that?

**Comment by pmoulon** [comment]
For the moment it's WIP. The code allow to perform feature tracking. So in order to do slam you  must need to connect relative pose estimation, absolute pose and then pose adjustment. All the piece are here, they just need to be connected.

**Comment by waqarsqureshi** [comment]
Ah, I thought ot is done.
